akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -548,17 +560,24 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
     val elem = connection.slot
 
     // Fast path
-    if ((connection.portState & (InReady | InFailed)) == InReady && (elem.asInstanceOf[AnyRef] ne Empty)) {
+    if ((connection.portState & (InReady | InFailed | InClosed)) == InReady && (elem.asInstanceOf[AnyRef] ne Empty)) {
       connection.slot = Empty
       elem.asInstanceOf[T]
     } else {
       // Slow path
       if (!isAvailable(in))
         throw new IllegalArgumentException(s"Cannot get element from already empty input port ($in)")
-      val failed = connection.slot.asInstanceOf[Failed]
-      val elem = failed.previousElem.asInstanceOf[T]
-      connection.slot = Failed(failed.ex, Empty)
-      elem
+
+      if ((connection.portState & (InReady | InFailed)) == (InReady | InFailed)) {
```

Comment:
a comment for each of these cases for the scenario?

Reply:
Yep, added here and elsewhere.